guides/keras_tuner/visualize_tuning.py

```diff
@@ -0,0 +1,180 @@
+visualizing the machine learning experiments.  It can monitor the losses and
+metrics during the model training and visualize the model architectures.
+Running KerasTuner with TensorBoard will give you additional features for
+visualizing hyperparameter tuning results using its HParams plugin.
+"""
+
+"""
+We will use a simple example of tuning a model for the MNIST image
+classification dataset to show how to use KerasTuner with TensorBoard.
+
+The first step is to download and format the data.
+"""
+
+import numpy as np
+from tensorflow import keras
+from tensorflow.keras import layers
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()
+# Normalize the pixel values to the range of [0, 1].
+x_train = x_train.astype("float32") / 255
+x_test = x_test.astype("float32") / 255
+# Add the channel dimension to the images.
+x_train = np.expand_dims(x_train, -1)
+x_test = np.expand_dims(x_test, -1)
+# Print the shapes of the data.
+print(x_train.shape)
+print(y_train.shape)
+print(x_test.shape)
+print(y_test.shape)
+
+"""
+Then, we write a `build_model` function to build the model with hyperparameters
+and return the model. The hyperparameters include the type of model to use
+(multi-layer perceptron or convolutional neural network), the number of layers,
+the number of units or filters, whether to use dropout.
+"""
+
+
+def build_model(hp):
+    num_classes = 10
+    input_shape = (28, 28, 1)
+    model = keras.Sequential()
+    model.add(keras.layers.InputLayer(input_shape=input_shape))
+    # Model type can be MLP or CNN.
+    model_type = hp.Choice("model_type", ["mlp", "cnn"])
+    if model_type == "mlp":
+        model.add(layers.Flatten())
+        # Number of layers of the MLP is a hyperparameter.
+        for i in range(hp.Int("mlp_layers", 1, 3)):
+            # Number of units of each layer are
+            # different hyperparameters with different names.
+            model.add(
+                layers.Dense(
+                    units=hp.Int("units_{i}".format(i=i), 32, 128, step=32),
+                    activation="relu",
+                )
+            )
+    else:
+        # Number of layers of the CNN is also a hyperparameter.
+        for i in range(hp.Int("cnn_layers", 1, 3)):
+            model.add(
+                layers.Conv2D(
+                    hp.Int("filters_{i}".format(i=i), 32, 128, step=32),
+                    kernel_size=(3, 3),
+                    activation="relu",
+                )
+            )
+            model.add(layers.MaxPooling2D(pool_size=(2, 2)))
+        model.add(layers.Flatten())
+
+    # A hyperparamter for whether to use dropout layer.
+    if hp.Boolean("dropout"):
+        model.add(layers.Dropout(0.5))
+
+    model.add(layers.Dense(units=num_classes, activation="softmax"))
+
+    # Compile the model.
+    model.compile(
+        loss="sparse_categorical_crossentropy",
+        metrics=["accuracy"],
+        optimizer=hp.Choice("optimizer", ["adam", "adadelta", "sgd"]),
```

Comment:
`adadelta` with default parameters is generally a bad choice (it needs a lr closer to 1 but by default uses 1e-3). You already have lots of hps in this example so tuning the optimizer isn't super necessary